Hammerfest.Server/Dns/HostsFile.cs

```diff
@@ -9,69 +8,82 @@ namespace Hammerfest.Server.Dns;
 /// </remarks>
 public static class HostsFile
 {
-    public static string? FindIpAddress(ISystemEnvironment env, string host)
+    /// <summary>
+    ///     Overrides the local hosts file.
+    /// </summary>
+    /// <param name="env">The <see cref="ISystemEnvironment" /> to modify hosts file in</param>
+    /// <param name="entriesToSet">
+    ///     The <see cref="IReadOnlyDictionary{TKey,TValue}" /> of new DNS entries, where the key is considered the host
+    ///     and the value is considered its IP address. Null as the entry value indicates that entry is to be deleted
+    /// </param>
+    /// <returns>
+    ///     The <see cref="Dictionary{TKey,TValue}" /> of initial DNS entries, where the key is considered the host
+    ///     and the value is considered its IP address.
+    /// </returns>
+    /// <remarks>
+    ///     Null as the entries' value indicates that the value is to be deleted.
+    ///     When new entry is added (not replaced) its resulting value in the returned dictionary is null. It is needed
+    ///     to reuse the method on the rollback.
+    /// </remarks>
+    public static Dictionary<string, string?> ModifyEntries(ISystemEnvironment env, IReadOnlyDictionary<string, string?>
+        entriesToSet)
     {
-        foreach (var line in File.ReadAllLines(env.HostsFilePath))
-        {
-            if (ParseLine(line) is (var ipAddress, var currentHost) && currentHost == host)
-                return ipAddress;
-        }
-
-        return null;
-    }
+        if (entriesToSet.Count == 0) return new Dictionary<string, string?>();
 
-    public static void AddEntry(ISystemEnvironment env, string ipAddress, string host)
-    {
-        var content = File.ReadAllText(env.HostsFilePath);
-        var result = new StringBuilder(content);
-        if (content.Length > 0 && !content.EndsWith('\n'))
-            result.Append('\n');
-        result.Append($"{ipAddress} {host}\r\n");
-
-        File.WriteAllText(env.HostsFilePath, result.ToString());
-    }
-
-    public static bool RemoveEntry(ISystemEnvironment env, string host)
-    {
+        var newEntries = new Dictionary<string, string?>(entriesToSet);
         var originalLines = File.ReadAllLines(env.HostsFilePath);
         var resultLines = new List<string>();
-        var removed = false;
+        var originalEntries = new Dictionary<string, string?>();
+
         foreach (var line in originalLines)
         {
-            if (ParseLine(line) is var (_, currentHost) && currentHost == host)
+            var parsedLine = ParseLine(line);
+            switch (parsedLine)
             {
-                removed = true;
+                case var (oldIp, host) when IsAlreadyPresent(host, oldIp):
+                    resultLines.Add($"{newEntries[host]} {host}");
```

Comment:
```suggestion
                    resultLines.Add(line);
```

To keep comments and formatting for the lines we don't wanna change.